book/src/dev/rfcs/drafts/0005-treestate.md

```diff
@@ -0,0 +1,203 @@
+computed treestate from the included transactions against the block header
+metadata (for Sapling and Orchard) or previously finalized state (for Sprout). This document
+describes how we compute and manage that data, assuming a finalized state
+service as described in the [State Updates RFC](https://zebra.zfnd.org/dev/rfcs/0005-state-updates.md).
+
+
+# Motivation
+[motivation]: #motivation
+
+Block validation requires checking that the treestate of the block (consisting
+of the note commitment tree and nullifier set) is consistent with the metadata
+we have in the block header (the root of the note commitment tree) or previously
+finalized state (for Sprout).
+
+
+# Definitions
+[definitions]: #definitions
+
+TODO: split up these definitions into common, Sprout, Sapling, and possibly Orchard sections
+
+Many terms used here are defined in the [Zcash Protocol Specification](https://zips.z.cash/protocol/protocol.pdf)
+
+**notes**: Represents a value bound to a shielded payment address (public key)
+which is spendable by the recipient who holds the spending key corresponding to
+a given shielded payment address.
+
+**nullifiers**: Revealed by `Spend` descriptions when its associated `Note` is spent.
+
+**nullifier set**: The set of unique `Nullifier`s revealed by any `Transaction`s
+within a `Block`. `Nullifier`s are enforced to be unique within a valid block chain
+by commiting to previous treestates in `Spend` descriptions, in order to prevent
+double-spends.
+
+**note commitments**: Pedersen commitment to the values consisting a `Note`. One
+should not be able to construct a `Note` from its commitment.
+
+**note commitment tree**: An incremental Merkle tree of fixed depth used to
+store `NoteCommitment`s that `JoinSplit` transfers or `Spend` transfers produce. It
+is used to express the existence of value and the capability to spend it. It is
+not the job of this tree to protect against double-spending, as it is
+append-only: that's what the `Nullifier` set is for.
+
+**note position**: The index of a `NoteCommitment` at the leafmost layer,
+counting leftmost to rightmost. The [position in the tree is determined by the
+order of transactions in the block](https://zips.z.cash/protocol/canopy.pdf#transactions).
+
+**root**: The layer 0 node of a Merkle tree.
+
+**anchor**: A Merkle tree root of a `NoteCommitment` tree. It uniquely
+identifies a `NoteCommitment` tree state given the assumed security properties
+of the Merkle tree’s hash function.  Since the `Nullifier` set is always updated
+together with the `NoteCommitment` tree, this also identifies a particular state
+of the associated `Nullifier` set.
+
+**spend descriptions**: A shielded Sapling transfer that spends a `Note`. Includes
+an anchor of some previous `Block`'s `NoteCommitment` tree.
+
+**output descriptions**: A shielded Sapling transfer that creates a
+`Note`. Includes the u-coordinate of the `NoteCommitment` itself.
```

Comment:
```suggestion
**spend descriptions**: A shielded Sapling transfer that spends a `Note`. Includes
an anchor of some previous `Block`'s `NoteCommitment` tree.

**output descriptions**: A shielded Sapling transfer that creates a
`Note`. Includes the u-coordinate of the `NoteCommitment` itself.

**action descriptions**: A shielded Orchard transfer that spends and/or creates a `Note`. 
Does not include an anchor, because that is encoded once in the `anchorOrchard` 
field of a V5 `Transaction`.


```